ath79: Add support for OpenMesh OM2P v1

ecsv opened the pull request:
This PR is based on the PR https://github.com/openwrt/openwrt/pull/3720

OpenMesh OM2P v1
================

Device specifications:
----------------------

* Qualcomm/Atheros AR7240 rev 2
* 350/350/175 MHz (CPU/DDR/AHB)
* 32 MB of RAM
* 16 MB of SPI NOR flash
  - 2x 7 MB available; but one of the 7 MB regions is the recovery image
* 2x 10/100 Mbps Ethernet
* 1T1R 2.4 GHz Wi-Fi
* 6x GPIO-LEDs (3x wifi, 2x ethernet, 1x power)
* 1x GPIO-button (reset)
* external h/w watchdog (enabled by default)
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 2x fast ethernet
  - eth0
    + 18-24V passive POE (mode B)
    + used as WAN interface
  - eth1
    + builtin switch port 4
    + used as LAN interface
* 12-24V 1A DC
* external antenna

The device itself requires the mtdparts from the uboot arguments to properly boot the flashed image and to support dual-boot (primary + recovery image). Unfortunately, the name of the mtd device in mtdparts is still using the legacy name "ar7240-nor0" which must be supplied using the Linux-specfic DT parameter linux,mtd-name to overwrite the generic name "spi0.0".

Flashing instructions:
----------------------

Various methods can be used to install the actual image on the flash.
Two easy ones are:

### ap51-flash

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be used to transfer the image to the u-boot when the device boots up.


### initramfs from TFTP

The serial console must be used to access the u-boot shell during bootup. It can then be used to first boot up the initramfs image from a TFTP server (here with the IP 192.168.1.21):

    setenv serverip 192.168.1.21
    setenv ipaddr 192.168.1.1
    tftpboot 0c00000 <filename-of-initramfs-kernel>.bin && bootm $fileaddr

The actual sysupgrade image can then be transferred (on the LAN port) to the device via

    scp <filename-of-squashfs-sysupgrade>.bin root@192.168.1.1:/tmp/

On the device, the sysupgrade must then be started using

    sysupgrade -n /tmp/<filename-of-squashfs-sysupgrade>.bin

ecsv committed:
ath79: Add support for OpenMesh OM2P v1

Device specifications:
======================

* Qualcomm/Atheros AR7240 rev 2
* 350/350/175 MHz (CPU/DDR/AHB)
* 32 MB of RAM
* 16 MB of SPI NOR flash
  - 2x 7 MB available; but one of the 7 MB regions is the recovery image
* 2x 10/100 Mbps Ethernet
* 1T1R 2.4 GHz Wi-Fi
* 6x GPIO-LEDs (3x wifi, 2x ethernet, 1x power)
* 1x GPIO-button (reset)
* external h/w watchdog (enabled by default)
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 2x fast ethernet
  - eth0
    + 18-24V passive POE (mode B)
    + used as WAN interface
  - eth1
    + builtin switch port 4
    + used as LAN interface
* 12-24V 1A DC
* external antenna

The device itself requires the mtdparts from the uboot arguments to
properly boot the flashed image and to support dual-boot (primary +
recovery image). Unfortunately, the name of the mtd device in mtdparts is
still using the legacy name "ar7240-nor0" which must be supplied using the
Linux-specfic DT parameter linux,mtd-name to overwrite the generic name
"spi0.0".

Flashing instructions:
======================

Various methods can be used to install the actual image on the flash.
Two easy ones are:

ap51-flash
----------

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
used to transfer the image to the u-boot when the device boots up.

initramfs from TFTP
-------------------

The serial console must be used to access the u-boot shell during bootup.
It can then be used to first boot up the initramfs image from a TFTP server
(here with the IP 192.168.1.21):

   setenv serverip 192.168.1.21
   setenv ipaddr 192.168.1.1
   tftpboot 0c00000 <filename-of-initramfs-kernel>.bin && bootm $fileaddr

The actual sysupgrade image can then be transferred (on the LAN port) to the
device via

  scp <filename-of-squashfs-sysupgrade>.bin root@192.168.1.1:/tmp/

On the device, the sysupgrade must then be started using

  sysupgrade -n /tmp/<filename-of-squashfs-sysupgrade>.bin

Signed-off-by: Sven Eckelmann <sven@narfation.org>

ecsv committed:
ath79: Add support for OpenMesh OM5P-AC v2

Device specifications:
======================

* Qualcomm/Atheros QCA9558 ver 1 rev 0
* 720/600/200 MHz (CPU/DDR/AHB)
* 128 MB of RAM
* 16 MB of SPI NOR flash
  - 2x 7 MB available; but one of the 7 MB regions is the recovery image
* 2T2R 2.4 GHz Wi-Fi (11n)
* 2T2R 5 GHz Wi-Fi (11ac)
* 4x GPIO-LEDs (3x wifi, 1x power)
* 1x GPIO-button (reset)
* external h/w watchdog (enabled by default))
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* TI tmp423 (package kmod-hwmon-tmp421) for temperature monitoring
* 2x ethernet
  - eth0
    + AR8035 ethernet PHY (RGMII)
    + 10/100/1000 Mbps Ethernet
    + 802.3af POE
    + used as LAN interface
  - eth1
    + AR8031 ethernet PHY (RGMII)
    + 10/100/1000 Mbps Ethernet
    + 18-24V passive POE (mode B)
    + used as WAN interface
* 12-24V 1A DC
* internal antennas

This device support is based on the partially working stub from commit
53c474abbdfe ("ath79: add new OF only target for QCA MIPS silicon").

Flashing instructions:
======================

Various methods can be used to install the actual image on the flash.
Two easy ones are:

ap51-flash
----------

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
used to transfer the image to the u-boot when the device boots up.

initramfs from TFTP
-------------------

The serial console must be used to access the u-boot shell during bootup.
It can then be used to first boot up the initramfs image from a TFTP server
(here with the IP 192.168.1.21):

   setenv serverip 192.168.1.21
   setenv ipaddr 192.168.1.1
   tftpboot 0c00000 <filename-of-initramfs-kernel>.bin && bootm $fileaddr

The actual sysupgrade image can then be transferred (on the LAN port) to the
device via

  scp <filename-of-squashfs-sysupgrade>.bin root@192.168.1.1:/tmp/

On the device, the sysupgrade must then be started using

  sysupgrade -n /tmp/<filename-of-squashfs-sysupgrade>.bin

Signed-off-by: Sven Eckelmann <sven@narfation.org>

adschm commented on package/boot/uboot-envtools/files/ath79:
```diff
@@ -83,7 +83,8 @@ netgear,wndr4300sw)
 plasmacloud,pa300|\
-plasmacloud,pa300e)
+plasmacloud,pa300e|\
+openmesh,om2p-v4)
```
Sorting, o is before p ...

ecsv replied:
done, will be pushed soon

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -237,6 +237,10 @@ pcs,cr3000)
 	ucidef_set_led_switch "lan3" "LAN3" "blue:lan3" "switch0" "0x10"
 	ucidef_set_led_switch "lan4" "LAN4" "blue:lan4" "switch0" "0x02"
 	;;
+openmesh,om2p-v4)
```
Sorting, o is before p ...

ecsv replied:
consider this done

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1526,6 +1526,24 @@ define Device/ocedo_ursus
+  IMAGE_SIZE := 7168k
+  BLOCKSIZE := 256k
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma | pad-to $$(BLOCKSIZE)
+  IMAGE/sysupgrade.bin := append-rootfs | pad-rootfs | openmesh-image ce_type=OM2P | append-metadata
```
Please wrap these lines to have max. 74 chars in one line.
Hanging lines with one tab for indent ...

ecsv replied:
Consider this done

adschm commented on target/linux/ath79/generic/base-files/lib/upgrade/openmesh.sh:
```diff
@@ -68,7 +68,8 @@ platform_do_upgrade_openmesh()
 	case "$(board_name)" in
-	openmesh,om2p-v4)
+	openmesh,om2p-v4|\
+	openmesh,om2p-hs-v4)
```
I wonder whether we really need another case here.

One could just make that another argument and then e.g. call `platform_do_upgrade_openmesh "$1" 3` in upgrade.sh

But please take this as an open thought, I'm not sure whether this applies to common behavior with these scripts and whether it's a good idea at all eventually ...

ecsv replied:
will be dropped completely

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1544,6 +1544,15 @@ define Device/openmesh_om2p-v4
+define Device/openmesh_om2p-hs-v4
+  $(Device/openmesh_om2p-common)
+  SOC := qca9533
+  DEVICE_MODEL := OM2P-HS
```
Since we overwrite it here anyway, I'd remove the DEVICE_MODEL from the common definition and put it directly into openmesh_om2p-v4 as well.

ecsv replied:
only 5 of 8 of devices overwrite it. but ok, it is definitely more consistent to have it always in the actual device definition.

consider this done

adschm commented on target/linux/ath79/dts/ar9341_openmesh_om2p-hs-v1.dts:
```diff
@@ -0,0 +1,8 @@
+#include "ar9341_openmesh_om2p-hs.dtsi"
+
+/ {
+	compatible = "openmesh,om2p-hs", "qca,ar9341";
```
-v1 missing here.

ecsv replied:
Good catch. This was introduced when I've changed the v1 devices to have the variant "v1" included - which was (unfortunately) after I returned the borrowed OM2P-HS v1. So this would definitely would have caused problems in the field.

Was now fixed

adschm commented on target/linux/ath79/dts/ar9341_openmesh_om2p-hs-v1.dts:
```diff
@@ -0,0 +1,8 @@
+
+/ {
+	compatible = "openmesh,om2p-hs", "qca,ar9341";
+	model = "OpenMesh OM2P-HS";
```
"v1" missing here.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -242,6 +242,10 @@ openmesh,om2p-hs-v4)
 	ucidef_set_led_netdev "wan" "WAN" "blue:wan" "eth0"
 	ucidef_set_led_switch "lan" "LAN" "blue:lan" "switch0" "0x02"
 	;;
+openmesh,om2p-hs-v1)
```
Case can be merged with
```
comfast,cf-ew72|\
telco,t1)
```

ecsv replied:
consider this consolidated

adschm commented on target/linux/ath79/dts/ar9341_openmesh_om2p-hs.dtsi:
```diff
@@ -0,0 +1,163 @@
+		compatible = "gpio-keys";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&jtag_disable_pins &led_lan_wan_blue_pin>;
```
What about
`ath79_gpio_output_select(OM2P_GPIO_BTN_RESET, AR934X_GPIO_OUT_GPIO);`
?

ecsv replied:
This line doesn't make a lot of sense. ath79_gpio_output_select switches the GPIO_OUT_FUNCTION registers for the output of an GPIO pin. But the OM2P_GPIO_BTN_RESET is an input. To switch the pin 1 to GPIO *input*, you would have to change 0x18040044- 0x1804006C (make sure that nothing uses the pin 1 for something else than GPIO). The relevant change to enable the input functionality for this button was actually the jtag_disable_pins because the pinmux has the pin 1 by default configured as JTAG TDI.

adschm commented on target/linux/ath79/dts/ar9341_openmesh_om2p-hs.dtsi:
```diff
@@ -0,0 +1,163 @@
+&eth0 {
+	status = "okay";
+
+	phy-handle = <&swphy4>;
```
Looks like ar71xx used physwap and phy0 here?

```
	/* GMAC0 is connected to the PHY0 of the internal switch */
	ath79_switch_data.phy4_mii_en = 1;
	ath79_switch_data.phy_poll_mask = BIT(0);
	ath79_eth0_data.phy_if_mode = PHY_INTERFACE_MODE_MII;
	ath79_eth0_data.phy_mask = BIT(0);
	ath79_eth0_data.mii_bus_dev = &ath79_mdio1_device.dev;
	ath79_register_eth(0);

	/* GMAC1 is connected to the internal switch */
	ath79_eth1_data.phy_if_mode = PHY_INTERFACE_MODE_GMII;
	ath79_register_eth(1);
```

adschm replied:
MAC addresses are swapped as well:

```
	u8 *mac1 = (u8 *)KSEG1ADDR(0x1ffc0000);
	u8 *mac2 = (u8 *)KSEG1ADDR(0x1ffc0000 + ETH_ALEN);
	ath79_init_mac(ath79_eth0_data.mac_addr, mac1, 0);
	ath79_init_mac(ath79_eth1_data.mac_addr, mac2, 0);

```

ecsv replied:
See https://github.com/openwrt/openwrt/pull/3710#pullrequestreview-558957526

adschm commented on target/linux/ath79/dts/ar9330_openmesh_om2p-lc.dtsi:
```diff
@@ -0,0 +1,151 @@
+&eth1 {
+	status = "okay";
+
+	mtd-mac-address = <&art 0x0>;
```
Again, MAC addresses inverted compared to ar71xx mach files.

ecsv replied:
You will find this a lot. Simply to avoid the other hack which you didn't like to reorder eth0 + eth1. And to (when possible) use gmac0 as WAN as promoted in the SoC specs.

adschm replied:
I'm not so sure anymore whether I still dislike the forced-order hack. I'd leave you to what you prefer there.

adschm commented on target/linux/ath79/dts/ar9330_openmesh_om2p-v2.dts:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "ar9330_openmesh_om2p-lc.dtsi"
```
With om2p-lc and om2p-v2 sharing this as the only devices on that soc, I'd call the DTSI ..._om2p.dtsi

ecsv replied:
Not sure about the name. But I also was not sure about the old name - so consider it done

adschm commented on target/linux/ath79/dts/ar7240_openmesh_om2p-v1.dts:
```diff
@@ -0,0 +1,160 @@
+&pcie {
+	status = "okay";
+
+	ath9k: wifi@0,0 {
```
DT label does not seem to be needed?

ecsv replied:
correct, it is only needed for mr600 - seems to be an artifact of an older test. consider it removed

adschm commented on target/linux/ath79/dts/ar7240_openmesh_om2p-v1.dts:
```diff
@@ -0,0 +1,160 @@
+			label = "green:wifi";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
```
Are the LEDs ordered somehow? If not, I'd recommend to order them by gpio number (only wifi_green needs to be moved) for this and the other devices (since you have to do other changes anyway) ...

ecsv replied:
No specific ordering scheme was used. I have now sorted it by gpio number for all devices

adschm commented:
I cannot review the upgrade script, somebody else will have to help for that.

ecsv commented:
The changes should now be pushed and available in all PRs.

Regarding the update script - this is basically the old openmesh.sh from ar71xx  - minus the the code which is now handled by the generic metadata and with the cleanups requested earlier for the dualboot_datachk.sh script

(btw. can please someone add the label target/ath79 to #3720?)

adschm commented:
> The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
> used to transfer the image to the u-boot when the device boots up.

I'd be happy about some manual instructions as well.

adschm commented on target/linux/ath79/generic/base-files/lib/upgrade/openmesh.sh:
```diff
@@ -0,0 +1,114 @@
+			[ "$cfg_opt" = "[$section]" ] && our_section=1 && continue
+			[ "$our_section" = "1" ] || continue
+
+			our_param=$(echo ${cfg_opt%%=*})
```
Is there a reason for this echo?

ecsv replied:
You would have to ask Gabor Juhos about this

adschm replied:
The only reason I can think of would be to remove newlines (and convert them to spaces).

But I can live with just keeping this for now.

ecsv replied:
This is rather unlikely because each entry of cfg_opt should not contain any newlines. I will have a look and may drop it

adschm replied:
Older OpenWrt is really crowded with useless echos. You actually removed several in the script already. So, probably, there really is no purpose after all.

But if you are unsure, as stated, I can live with keeping this one for now.

ecsv replied:
Consider it dropped

ecsv commented:
> > The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
> > used to transfer the image to the u-boot when the device boots up.
> 
> I'd be happy about some manual instructions as well.

I don't have any manual instructions

adschm commented on target/linux/ath79/generic/base-files/lib/upgrade/openmesh.sh:
```diff
@@ -0,0 +1,114 @@
+
+	# write image parts
+	mtd -q erase inactive
+	dd if="$img_path" bs=1 skip=$(($data_offset + 0x$cfg_size + 0x$kernel_size)) count=$((0x$rootfs_size)) 2>&- | mtd -n -p $kernel_length $restore_backup write - $PART_NAME
```
bs=1 will be _very_ slow here? One should consider to improve that. But probably in a separate patch ...

ecsv replied:
Yes, it is rather slow. Increases the write time of the default image (without feeds) on an OM2P-HS v3 from roughly 30s to 120s. I will continue the discussion in the other comment

adschm commented on target/linux/ath79/generic/base-files/lib/upgrade/openmesh.sh:
```diff
@@ -0,0 +1,114 @@
+	local cfg_content=$(dd if="$img_path" bs=1 skip=$data_offset count=$((0x$cfg_size)) 2>/dev/null)
+
+	local kernel_size=$(dd if="$img_path" bs=2 skip=71 count=4 2>/dev/null)
+	local kernel_length=$((0x$kernel_size))
```
kernel_size and kernel_length are both used below. Either just assign *_size here and use 0x$var for everything below, or already create the 0x$var version for _all_ relevant variables here already (cfg_size, kernel_size, rootfs_size).

adschm replied:
The latter might actually improve readability a lot ...

ecsv replied:
consider it done (will be pushed after a coupe of tests)

adschm commented on target/linux/ath79/generic/base-files/lib/upgrade/openmesh.sh:
```diff
@@ -0,0 +1,114 @@
+
+	# write image parts
+	mtd -q erase inactive
+	dd if="$img_path" bs=1 skip=$(($data_offset + 0x$cfg_size + 0x$kernel_size)) count=$((0x$rootfs_size)) 2>&- | mtd -n -p $kernel_length $restore_backup write - $PART_NAME
```
We assume that kernel/rootfs is always aligned to blocksize now? Why hasn't this been done for ar71xx?

```
	if [ "$((0x$kernel_size % $block_size))" = "0" ]
		then
			kernel_bsize=$(echo $((0x$kernel_size)))
		else
			kernel_bsize=$((0x$kernel_size + ($block_size - (0x$kernel_size % $block_size))))
	fi

```

ecsv replied:
Because I have no time machine

adschm commented on target/linux/ath79/generic/base-files/lib/upgrade/openmesh.sh:
```diff
@@ -0,0 +1,114 @@
+	[ "$inactive_mtd" = "$primary_kernel_mtd" ] || next_boot_part="2"
+
+	local cfg_size=$(dd if="$img_path" bs=2 skip=35 count=4 2>/dev/null)
+	local cfg_content=$(dd if="$img_path" bs=1 skip=$data_offset count=$((0x$cfg_size)) 2>/dev/null)
```
That's also a candidate for swapping bs and count (needs additional skip-bytes then, of course).

ecsv replied:
Please don't swap bs and count. Will end up easily with copy problems on busybox with MIPS. I've tried it for the sysupgrade-tar on PA300 and ended up with half "dd"ed images. Most likely memory problems which caused the dd to stop and thus the tar to have write errors to the pipe.

adschm replied:
We are using skip_bytes and seek_bytes in various places, e.g. for caldata extraction (ath9k/ath10k). Why would this be different here?

I agree that dd-ing an entire kernel/rootfs in one chunk is not really a good idea. But doing so with bs=1 isn't desirable either. And talking about cfg_content here, this is reading (not writing) and probably smaller anyway.

But anyway, this is a separate discussion, I would definitely not touch bs/count for this commit/PR.

ecsv replied:
Ok, just checked the source code and you seem to be right. My statement was wrong that this isn't compiled in by default in openwrt.

adschm replied:
And since we assume blocksize-multiples for kernel (and rootfs as well?) now anyway, we could relatively simple work with bs=1k or similar there (even something like bs=64 would probably be an enormous speed improvement without jeopardizing the whole process too much).

ecsv replied:
Just to repeat the current results with bs=1024 from the other comment thread: Decreases the write time of the default image (without feeds) on an OM2P-HS v3 from roughly 120s to 30s. The erase + some overhead add in both cases roughly another 50s

adschm replied:
Thanks for testing, that's really a speedup, isn't it ;-)

But note that even in case I convinced you, this (bs vs. count) should be a separate patch, so we can revert/alter it easily later.

ecsv replied:
Problem is only that it doesn't work for the rootfs because the rootfs is (64k or 256k) * number of blocks + 4 bytes. Here we would need the count_bytes or an ugly hack to append the last "block".

adschm replied:
Theoretically, one could use bs=4 and would already be faster, but that's not really less ugly.

I wonder whether we can just get around that with `conv=sync` or similar. A few additional zeros should not hurt, and make everything easier.

I've never dealt with count_bytes so far.

ecsv replied:
wouldn't be a couple of zeros because the metadata follows after the rootfs

ecsv replied:
I have now added a minor change (with large boilerplate) which just rounds up the number of 1024 byte blocks for rootfs - which means that it write the "End of filesystem" marker to flash + in some circumstances bogus sysupgrade metadata.

These "circumstances" would be when no config would be saved during a sysupgrade.

adschm commented:
> Regarding the update script - this is basically the old openmesh.sh from ar71xx - minus the the code which is now handled by the generic metadata and with the cleanups requested earlier for the dualboot_datachk.sh script

Nice improvement of the original code.

Left some comments there, should be fine based on the comparison with the previous code, assuming you tested it. I'm really no expert with sysupgrade, but in worst case it cannot do much harm either.

adschm commented:
> I don't have any manual instructions. In theory, you could boot the initramfs image up like this:

I was just assuming that the tool is doing something that one could also do manually instead.

adschm commented:
BTW, if you apply the changes apart from the bs/count discussion to this PR already, I'd relatively quickly merge at least the first two devices. That might make everything else easier.

ecsv commented:
Update everything except the stuff in the controversial conversations (which are still marked as "open conversations")

adschm commented:
I've pushed the first two devices and the bs/count patch for now.

For the remaining devices, I'm still confused by the swapped MAC addresses...

adschm commented:
And what about simply adding 'pad-to $$$$(BLOCK_SIZE)' before append-metadata?

ecsv commented:
> For the remaining devices, I'm still confused by the swapped MAC addresses...

Let me try to explain it via OM2P-HS: They were using using P0 and P4 (called swphy0 and swphy4 in the DTS). For some reason unknown to me, the gmac0 (eth0 under Linux) was connected to P0 (LAN) and had the +0 mac address. And gmac1 (eth1 under Linux) had the P4 (WAN) and the +1 mac address.

Now with ath79, the eth1 under Linux is gmac0. And eth0 under Linux is gmac1. So I've swapped the mac addresses too and the P0+P4. The mac addresses are therefore still the same for P0 and P4 - just the attached gmac's were swapped.

| target        | P0 (LAN)                     | P4 (WAN)                     |
| ------------- |------------------------------|------------------------------|
| ar71xx        | gmac0 aka Linux eth0, +0 mac | gmac1 aka Linux eth1, +1 mac |
| ath79         | gmac1 aka Linux eth0, +0 mac | gmac0 aka Linux eth1, +1 mac |

And gmac0 is also the one which is not connected to the switch (and has a lot of other nice HW blocks) and is therefore a good match for a WAN interface (under ath79).

I hope I didn't confuse myself during the explanation and didn't write down complete stupid stuff.

adschm commented:
> And gmac0 is also the one which is not connected to the switch (and has a lot of other nice HW blocks) and is therefore a good match for a WAN interface (under ath79).

And that's the reason why you don't use the PHY_SWAP in ath79 as it's done in ar71xx? 

`ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_SW_PHY_SWAP);`

This is not swapped intentionally as you prefer a different gmac for WAN?

ecsv commented:
> And that's the reason why you don't use the PHY_SWAP in ath79 as it's done in ar71xx?

Exactly. And everything falls into place when not doing the P0<->P4 swap.

> This is not swapped intentionally as you prefer a different gmac for WAN?

Actually, the SoC and the SoC integration in ath79 prefers gmac0 for WAN. And thus I've started to prefer gmac0 for WAN too because everything else pushes in this direction.


Btw. this brings me to a point where you maybe have more ideas: The gmac1 is for ar9344 connected to the switch with an fixed-link. The ag71xx_phy will detect this based on the DT and handle this like a 1000 MB/s fixed link. This means that the interface will not switch its carrier state when the (only) PHY on the switch changes its carrier state.

Are there already other devices in ath79 with only one exposed (non-CPU) switch port which propagates the switch port state directly to the ethernet device under Linux?

adschm commented:
Okay, I picked the OM2P HS bunch.

Next obstacle is the LC, as this one does not have a phy swap in ar71xx, but the MAC addresses are still swapped in ath79, with 01_leds/02_network still assigned the same way?

ecsv commented:
> Next obstacle is the LC, as this one does not have a phy swap in ar71xx, but the MAC addresses are still swapped in ath79, with 01_leds/02_network still assigned the same way?

Short summary: Big mess in ar71xx.

See following snippet from my OM2P v2 (ar71xx) with the LAN port connected:

    root@OpenWrt:/# ip link
    1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
        link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel master br-lan state DOWN qlen 1000
        link/ether ac:86:74:00:37:c8 brd ff:ff:ff:ff:ff:ff
    3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
        link/ether ac:86:74:00:37:c9 brd ff:ff:ff:ff:ff:ff
    4: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
        link/ether ac:86:74:00:37:ca brd ff:ff:ff:ff:ff:ff
    5: br-lan: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN qlen 1000
        link/ether ac:86:74:00:37:c8 brd ff:ff:ff:ff:ff:ff

You might notice that br-lan is down - even when it should actually be up. But WAN is up (which shouldn't be the case).

So you could say (for example): What is missing in ar71xx is a PHY and MAC address swap ;)

adschm commented:
I will merge LC, v2 and the OM5P from the other PR soon. I'm not so familiar with the qca955x/rgmii stuff, so the remaining devices might take a little longer and/or be taken by others.

Can you comment a little more extensively on the boot changes for v1 in this PR, specifically with respect to what you refer to with "renamed devices"?

ecsv commented:
> Can you comment a little more extensively on the boot changes for v1 in this PR, specifically with respect to what you refer to with "renamed devices"?

Yes, I will extend the description in the commit. Just have to make a dive in the box of borrowed devices to provide the correct names.

I will ping afterwards.

ecsv commented:
@adschm I had another look whether it is possible somehow with the SPI devicetree bindings to overwrite the SPI identifier but was not (yet) successful. I have therefore just changed the commit message to provide more insight.

ecsv commented:
Just found out that there is already a patch for pistachio which supports this (401-mtd-nor-support-mtd-name-from-device-tree.patch). I would propose to import this to ath79 instead of forcing the user to change the u-boot.

ecsv committed:
ath79: Allow to set spi-nor mtd name via devicetree

On ar71xx, it was possible to overwrite the name of the spi-nor mtd device
identifier using the flash_platform_data which each mach-*.c could adjust
for its devices. A similar feature was introduced for mtd-physmap in
devicetree's. The property linux,mtd-name can be used to set the name and
provide a stable identifier for mtdpart from the bootloader.

But this feature is not yet available upstream for spi-nor devices which
also might receive their partition layout from the bootloader. But the
OpenWrt pistachio support for this property can simply be imported into
ath79 to gain this support.

Signed-off-by: Sven Eckelmann <sven@narfation.org>

ecsv commented:
@adschm I have now pushed a version which doesn't need any u-boot changes - thus it allows a seamingless upgrade from ar71xx to ath79.

The old version (which required an u-boot change) can still be found at https://github.com/ecsv/openwrt/tree/om2p-change-bootloader-mtd-name

CodeFetch commented:
@ecsv Is the recovery image actually 7 MB in size? If not you could still use the unused space for OpenWrt with mtdsplit...

ecsv commented:
@CodeFetch  The recovery "image" is taking up the rest of the space

hauke commented on target/linux/ath79/patches-5.10/401-mtd-nor-support-mtd-name-from-device-tree.patch:
```diff
@@ -0,0 +1,51 @@
+From f32bc2aa01edcba2f2ed5db151cf183eac9ef919 Mon Sep 17 00:00:00 2001
```
Are you taking the bootargs from U-Boot and do not overwrite them in the OpenWrt device tree?
If you nee the patch which adds the mtd name because U-Boot uses a specific mtd name in the bootargs, then please add this to the commit description.

ecsv replied:
Yes, I have now added an extra paragraph about this to the commit message.

hauke commented:
Thank you for your patches, I applied them to master in 7f074fe1a79f27d30560c35778d963bc520e9a3e and 8143709c90c3aa0a84e14cf4a9dfbcb365b0d966.
